securedrop/encryption.py

```diff
@@ -134,63 +102,48 @@ def get_source_key_fingerprint(self, source_filesystem_id: str) -> str:
         return source_key_fingerprint
 
     def encrypt_source_message(self, message_in: str, encrypted_message_path_out: Path) -> None:
-        message_as_stream = StringIO(message_in)
-        self._encrypt(
+        redwood.encrypt_message(
             # A submission is only encrypted for the journalist key
-            using_keys_with_fingerprints=[self._journalist_key_fingerprint],
-            plaintext_in=message_as_stream,
-            ciphertext_path_out=encrypted_message_path_out,
+            recipients=[self.get_journalist_public_key()],
+            plaintext=message_in,
+            destination=encrypted_message_path_out,
         )
 
-    def encrypt_source_file(self, file_in: typing.IO, encrypted_file_path_out: Path) -> None:
-        self._encrypt(
+    def encrypt_source_file(self, file_in: BinaryIO, encrypted_file_path_out: Path) -> None:
+        redwood.encrypt_stream(
             # A submission is only encrypted for the journalist key
-            using_keys_with_fingerprints=[self._journalist_key_fingerprint],
-            plaintext_in=file_in,
-            ciphertext_path_out=encrypted_file_path_out,
+            recipients=[self.get_journalist_public_key()],
+            plaintext=file_in,
+            destination=encrypted_file_path_out,
         )
 
     def encrypt_journalist_reply(
-        self, for_source_with_filesystem_id: str, reply_in: str, encrypted_reply_path_out: Path
+        self, source: "Source", reply_in: str, encrypted_reply_path_out: Path
     ) -> None:
-        source_key_fingerprint = self.get_source_key_fingerprint(for_source_with_filesystem_id)
-        reply_as_stream = StringIO(reply_in)
-        self._encrypt(
+        redwood.encrypt_message(
             # A reply is encrypted for both the journalist key and the source key
-            using_keys_with_fingerprints=[source_key_fingerprint, self._journalist_key_fingerprint],
-            plaintext_in=reply_as_stream,
-            ciphertext_path_out=encrypted_reply_path_out,
+            recipients=[source.public_key, self.get_journalist_public_key()],
+            plaintext=reply_in,
+            destination=encrypted_reply_path_out,
         )
 
-    def decrypt_journalist_reply(self, for_source_user: "SourceUser", ciphertext_in: bytes) -> str:
+    def decrypt_journalist_reply(
+        self, for_source_user: "SourceUser", for_source: "Source", ciphertext_in: bytes
```

Comment:
I went back and forth on this a few times so I'm glad you're raising this because I'm not really satisfied with the current implementation. I didn't want to call `get_db_record()` since it seems wrong that the EncryptionManager is making database queries - and we already have the `Source` on hand. The issue of for_source_user/for_source not matching is bad, which I hadn't really thought of.

Reply:
I'm going to switch to just taking the SourceUser and calling get_db_record() with a TODO comment about not making DB queries.

Reply:
One way to solve this coupling would be with a decorator, e.g.:

```diff
--- a/securedrop/encryption.py
+++ b/securedrop/encryption.py
@@ -132,9 +132,10 @@ class EncryptionManager:
             destination=encrypted_reply_path_out,
         )
 
-    def decrypt_journalist_reply(self, for_source_user: "SourceUser", ciphertext_in: bytes) -> str:
-        # TODO: we shouldn't be making database queries here
-        for_source = for_source_user.get_db_record()
+    @with_source_from_source_user
+    def decrypt_journalist_reply(
+        self, for_source_user: "SourceUser", for_source: "Source", ciphertext_in: bytes
+    ) -> str:
         if for_source.pgp_secret_key is not None:
             return redwood.decrypt(
                 ciphertext_in,
```

That would let `with_source_from_source_user` be the glue between `decrypt_journalist_reply()`, which needs the `Source` but which we don't want to query the database, and its caller, which we don't want to be able pass an unmatched `SourceUser`/`Source` pair.

Reply:
I don't think that really helps very much since in the only call site, we already have the `Source` object on hand, so really it should just be passing it through.

Maybe we just add a manual assertion that the two source objects point to the same filesystem_id?

(The underlying issue that trips me up is that SourceUser doesn't contain a Source object, instead loading it on-demand for unclear (to me) reasons: "We don't store the actual record to force a refresh". Don't think we need to solve that now though...)

Reply:
Right.  I think it would need to be an explicit `if...: raise` (rather than `assert`), but I'd be happy with that approach, @legoktm.

@nabla-c0d3, the `SourceUser.get_db_record()` originates in your 203a873b1ee16f6b590aa31dced9bf00eaccd632: can you refresh (as it were :-) our understanding of why this fetch is best done on demand?

Reply:
@cfm It was to ensure an up-to-date record is being used. Otherwise, if we just attach the `models.Source` to the `SourceUser`, the DB record can get stale if it gets updated some time later. I think SQLAlchemy might try to auto-fetch/update it but it will raise an exception if there is no DB session currently opened.

I'm not super happy with my `get_db_record()` approach tho, as it fetches data from the DB without explicitly using a DB session, which, in my opinion, is an anti-pattern (hidden "dependencies" / magical behavior, and will throw an exception if no DB session is currently open). 

So overall I think it's not "worse" to just attach the `models.Source` to the `SourceUser`, as both approaches end up having the same problem. So I would be okay with that change.

Here's an example of the problem:

```
# Open a DB session, fetch the Source record, and create a SourceUser
with get_db_session() as db_session:
  source = db_session.query(models.Source).filter_by(filesystem_id=f_id).one()
  source_user = SourceUser(source)
  print(source_user.source)  #  Will work fine
  print(source_user.source.submissions)  #  Will work fine

# DB session is now closed
print(source_user.source)  #  Will work fine
print(source_user.source.submissions)  #  Will throw an exception
```

I think we can worry about it once it's a real problem, and like I said, my `get_db_record()` approach does not address this issue either.

Reply:
Thanks, @nabla-c0d3; I agree with the problem.  :-)  If we're going to have this anti-pattern either way, and it's not *enough* of a problem to be worth solving architecturally, then I'll defer to @legoktm on which anti-pattern we accept here.

Reply:
> the DB record can get stale if it gets updated some time later. I think SQLAlchemy might try to auto-fetch/update it but it will raise an exception if there is no DB session currently opened.

My understanding is that `models.Source` is just a reference to the actual, unique object in SQLAlchemy's identity map that corresponds to the row, and if something makes changes, the object will also be updated. And if there's no session, it should error.

So...my preference would be go to back to passing both source_user + source in, adding a check to verify they're the same, and creating an issue + leaving a TODO on get_db_record() to figure out what to do with it.

Reply:
Based on this discussion, that would be my preference too, @legoktm.

Reply:
Yeah I buy that. Ok, flip-flopping once more :)

Reply:
I filed https://github.com/freedomofpress/securedrop/issues/6934 summarizing the `get_db_record()` part of this convo.